### PR TITLE
Add text tooltip for TabBar & TabContainer

### LINKED ISSUE #9252: Add text tooltip for TabBar & TabContainer
### Describe the project you are working on

Data editing tool.

### Describe the problem or limitation you are having in your project

Currently, `TabBar` does not show text tooltips for tabs. It would be useful to be able to set custom text tooltips.
изображение

Also it would be useful to display a text tooltip with the tab title if the max tab width limit is set and the title does not fit.
изображение

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Need to override the `get_tooltip` method and add two new `set_tab_tooltip` & `get_tab_tooltip` methods to `TabBar`.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```c++
String TabBar::get_tooltip(const Point2 &p_pos) const {
	for (int i = 0; i < tabs.size(); i++) {
		if (tabs[i].hidden) {
			continue;
		}

		if (get_tab_rect(i).has_point(p_pos)) {
			if (tabs[i].tooltip.is_empty()) {
				return tabs[i].text;
			}

			return tabs[i].tooltip;
		}
	}

	return Control::get_tooltip(p_pos);
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes. It is possible to inherit from `TabBar` and override the `_get_tooltip` method.

### Is there a reason why this should be core and not an add-on in the asset library?

No.

### PR DESCRIPTION
**Pull request by 4d49**
Closes https://github.com/godotengine/godot-proposals/issues/9252

### CODE REVIEW

**Review by YeldhamDev — APPROVED** [review]
LGTM

**Review by akien-mga — APPROVED** [review]
Code looks pretty clean, and the feature makes sense. I'm surprised we didn't have it yet :D

File: scene/gui/tab_bar.cpp [hunk 1]
```diff
@@ -323,6 +323,24 @@ void TabBar::gui_input(const Ref<InputEvent> &p_event) {
...
 
+String TabBar::get_tooltip(const Point2 &p_pos) const {
+	for (int i = 0; i < tabs.size(); i++) {
+		if (tabs[i].hidden) {
+			continue;
+		}
+
+		if (get_tab_rect(i).has_point(p_pos)) {
+			if (tabs[i].tooltip.is_empty()) {
+				return tabs[i].text;
```
Comment (kitbdev) [inline scene/gui/tab_bar.cpp:334]: The fact that the tab title is used if it the tooltip is empty should probably be mentioned in the docs. 
Also, the only way to not have a tooltip is to set it to whitespace.

**Comment by KoBeWi** [comment]
The tooltip overlaps preview of scene tabs:
image

**Comment by Mickeon** [comment]
I remember there was some discussion ongoing in the past. Something about tooltips that only repeat the hovered subject being very annoying and superfluous. I wouldn't want to contribute to that by default as by this PR does (and as by KoBeWi's screenshot shows).

**Comment by YeldhamDev** [comment]
> I remember there was some discussion going in the past. Something about tooltips that only repeat the hovered subject being very annoying and superfluous.

I think those tooltips have a place when the tab's text is truncated, but yeah, it currently just shows it no matter what.

**Comment by ryevdokimov** [comment]
+1 to Mickeon's point. Tooltips will hijack 'ui_cancel' to close the tooltip. Select a node and then have your cursor over a tool mode button w/ the tooltip visible, then press escape (the default), and you'll see what I mean. 

I don't have an opinion on that specific behavior, but I do believe that it should be minimized, by not having tooltips showing redundant information scattered around the editor.

**Comment by Mickeon** [comment]
Point is, regarding this specific PR, the default behavior should be no tooltip at all (in my opinion)

File: scene/gui/tab_bar.cpp [hunk 2]
```diff
@@ -998,7 +1021,7 @@ void TabBar::_update_cache(bool p_update_hover) {
 		tabs.write[i].size_text = Math::ceil(tabs[i].text_buf->get_size().x);
 		tabs.write[i].size_cache = get_tab_width(i);
 
-		if (max_width > 0 && tabs[i].size_cache > max_width) {
+		if (tabs.write[i].truncated = max_width > 0 && tabs[i].size_cache > max_width) {
```
Comment (AThousandShips) [inline scene/gui/tab_bar.cpp:1024]:
```suggestion
		if (tabs.write[i].truncated == max_width > 0 && tabs[i].size_cache > max_width) {
```

File: scene/gui/tab_bar.cpp [hunk 3]
```diff
@@ -998,7 +1021,7 @@ void TabBar::_update_cache(bool p_update_hover) {
 		tabs.write[i].size_text = Math::ceil(tabs[i].text_buf->get_size().x);
 		tabs.write[i].size_cache = get_tab_width(i);
 
-		if (max_width > 0 && tabs[i].size_cache > max_width) {
+		if ((tabs.write[i].truncated = max_width > 0 && tabs[i].size_cache > max_width)) {
```
Comment (AThousandShips) [inline scene/gui/tab_bar.cpp:1024]: Is this a typo? If not it should be:
```suggestion
		tabs.write[i].truncated = max_width > 0;
		if (max_width > 0 && tabs[i].size_cache > max_width) {
```

Comment (4d49) [reply scene/gui/tab_bar.cpp:1024]: Yes sorry, the compiler on my computer didn't give any errors :(

**Comment by 4d49** [comment]
I reworked the code and now tooltips shows tab title only if text is truncated.

**Comment by Mickeon** [comment]
I think that makes sense, it's desirable and expected.
Are there cases in the editor where the tab title would appear truncated? Struggling to think of any right now, most tabs are not afraid to display their title in full.

File: scene/gui/tab_bar.cpp (hunk 3 above)
Comment (AThousandShips) [reply scene/gui/tab_bar.cpp:1024]: So should it be?
```suggestion
		if (tabs.write[i].truncated == max_width > 0 && tabs[i].size_cache > max_width) {
```
What is the intention of this code?

**Comment by KoBeWi** [comment]
I'm not even sure how to make a tab truncated, as tabs simply disappear if they have no space.

**Comment by YeldhamDev** [comment]
`max_tab_width` will limit how big a tab can get, and it's used in the editor for scene files with really long names:
image

**Review by AThousandShips — DISMISSED** [review]
The above code needs figuring out

File: scene/gui/tab_bar.cpp (hunk 3 above)
Comment (4d49) [reply scene/gui/tab_bar.cpp:1024]: > So should it be?

I meant the equivalent of this code here:
```C++
tabs.write[i].truncated = max_width > 0 && tabs[i].size_cache > max_width;
if (tabs[i].truncated) {
}
```

File: doc/classes/TabContainer.xml [hunk 4]
```diff
@@ -173,6 +180,14 @@
 				Sets a custom title for the tab at index [param tab_idx] (tab titles default to the name of the indexed child node). Set it back to the child's name to make the tab default to it again.
 			</description>
 		</method>
+		<method name="set_tab_tooltip">
+			<return type="void" />
+			<param index="0" name="tab_idx" type="int" />
+			<param index="1" name="tooltip" type="String" />
+			<description>
+				Sets a custom tooltip text for tab at index [param tab_idx].
```
Comment (kitbdev) [inline doc/classes/TabContainer.xml:188]:
```suggestion
				Sets a custom tooltip text for tab at index [param tab_idx].
				[b]Note:[/b] By default, if the [param tooltip] is empty and the tab text is truncated (not all characters fit into the tab), the title will be displayed as a tooltip. To hide the tooltip, assign [code]" "[/code] as the [param tooltip] text.
```

This should be the same as in the TabBar docs.

File: scene/gui/tab_bar.cpp (hunk 3 above)
Comment (kitbdev) [inline scene/gui/tab_bar.cpp:1024]:
```suggestion
		tabs.write[i].truncated = max_width > 0 && tabs[i].size_cache > max_width;
		if (tabs[i].truncated) {
```

This should be split into two lines as mentioned above. https://github.com/godotengine/godot/pull/89247#discussion_r1518476658

**Review by timothyqiu — DISMISSED** [review]
The tooltip text should be carried over in `TabContainer::move_tab_from_tab_container()` so that dragging across different containers work.

File: doc/classes/TabContainer.xml [hunk 5]
```diff
@@ -173,6 +180,15 @@
...
+				[b]Note:[/b] By default, if the [param tooltip] is empty and the tab text is truncated (not all characters fit into the tab), the title will be displayed as a tooltip. To hide the tooltip, assign [code]" "[/code] as the [param tooltip] text.
```
Comment (AThousandShips) [inline doc/classes/TabContainer.xml:189]:
```suggestion
				[b]Note:[/b] By default, if the [param tooltip] is empty and the tab text is truncated (not all characters fit into the tab), the title will be displayed as a tooltip. To hide the tooltip, assign [code]""[/code] as the [param tooltip] text.
```

File: doc/classes/TabBar.xml [hunk 6]
```diff
@@ -224,6 +231,15 @@
 				Sets a [param title] for the tab at index [param tab_idx].
 			</description>
 		</method>
+		<method name="set_tab_tooltip">
+			<return type="void" />
+			<param index="0" name="tab_idx" type="int" />
+			<param index="1" name="tooltip" type="String" />
+			<description>
+				Sets a [param tooltip] for tab at index [param tab_idx].
+				[b]Note:[/b] By default, if the [param tooltip] is empty and the tab text is truncated (not all characters fit into the tab), the title will be displayed as a tooltip. To hide the tooltip, assign [code]" "[/code] as the [param tooltip] text.
```
[the remark this answers is not in the document]
Comment (timothyqiu) [reply doc/classes/TabBar.xml]: Using an empty string shows the default truncation tooltip. At least one space should be used.

But technically tooltip is hidden only when `get_tooltip()` returns an empty string. Tooltip striping edges is an implementation detail. This way of hiding tooltip feels hacky.

[the remark this answers is not in the document]
Comment (4d49) [reply doc/classes/TabBar.xml]: My problem, I'm sorry. Of course we need a space there! It's just that the commit was a long time ago and I've already forgotten... Space is needed to hide the tooltip if the tab is truncated. @kitbdev requested a mention above.

**Review by timothyqiu — APPROVED** [review]
Needs a rebase, otherwise looks good.

**Comment by 4d49** [comment]
@timothyqiu done!

**Comment by Lexpeartha** [comment]
> Point is, regarding this specific PR, the default behavior should be no tooltip at all (in my opinion)

Also for the record, it doesn't seem possible to disable this behavior if you're using `TabContainer`  since `TabBar` is built-in so you can't do anything about it. Not sure if this is desired behavior though

**Pull request merged**